### PR TITLE
feat(compass-workspace): Update workspace tabs to Leafy Green in compass-workspace COMPASS-5518

### PR DESCRIPTION
**Pull request by Anemy**
COMPASS-5518

This PR updates the workspace tabs in Compass. It adds a new package `compass-workspace` with the tabs in it. These are now required from `compass-collection` where they were previously. The plan with `compass-workspace` is to have a higher level concept for tabs so they aren't bound to a collection. Please share any thoughts y'all might have on this.

I'll be syncing with Claudia for a last pass on styles before merging so I can bring any feedback there - let me know if you try it out and think of anything. I think one thing that might change is we'll add a minimum width for the tabs so they don't resize then the title grows or shrinks a bit.

These updates to tabs should bring more accessibility - if y'all see any ways we could improve please share. Here's a resource with more information: https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html

`env COMPASS_LG_DARKMODE="true" npm start` for component darkmode.
Screen Shot 2022-03-16 at 12 40 54 AM

https://user-images.githubusercontent.com/1791149/158519730-f1cc35bb-094a-4ec6-89c7-124f7e82c29f.mp4

*Scroll now overlays, styling in video a little out of date:

https://user-images.githubusercontent.com/1791149/158519733-eea4cb08-2400-43c1-a77c-8330ad640046.mp4

Has full borders now now, styling in video a little out of date:

https://user-images.githubusercontent.com/1791149/158519735-0c6d8a1c-748b-4f01-8732-22cd4c2fac24.mp4

### CODE REVIEW

File: packages/compass-components/src/components/workspace-container.tsx
Comment (Anemy) [inline packages/compass-components/src/components/workspace-container.tsx:38]: Unrelated drive by type improvement I accidentally scooped in this branch 😅

File: packages/compass-workspace/src/components/tab.tsx
Comment (gribnoysup) [inline packages/compass-workspace/src/components/tab.tsx:54]: This doesn't look like it's showing up or do anything, it is a leftover?

File: packages/compass-workspace/src/components/tab.tsx
Comment (gribnoysup) [inline packages/compass-workspace/src/components/tab.tsx:302]: It's not bad to have this role here, but you don't really need it for nodes without content

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (gribnoysup) [inline packages/compass-workspace/src/components/workspace-tabs.tsx:233]: Does it makes sense to have this as part of the component interface? I know that it was like that before, but aren't we only interested in collection type here (like view or timeseries) similar to how it is done in the sidebar navigation

File: packages/compass-workspace/src/components/tab.tsx
Comment (gribnoysup) [inline packages/compass-workspace/src/components/tab.tsx:241]: Because tabIndex never changes on these elements when using keyboard to navigate them with VO enabled you never get informed that the tab is changed

**Comment by gribnoysup** [comment]
> The plan with compass-workspace is to have a higher level concept for tabs so they aren't bound to a collection. Please share any thoughts y'all might have on this.

My feeling is that it makes sense now because tabs are used by collection plugin and nothing else, but feels like when tabs are really "global" we will have a separate package with 3 components that will be used in compass-home and nowhere else

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (mcasimir) [reply packages/compass-workspace/src/components/workspace-tabs.tsx:233]: To go one step forward, is there a way to completely split the code that "knows" about the domain of what the tab content is doing from the one that renders / moves and locates the tabs?

I would imagine something like this (ideally and very pseudo-codeish):

```
<WorkspaceTabs tabs={[
    {
       id: 'my-tab-123',
       header: <CollectionTabHeader ns="my-db.my-coll" type="timeseries"> 
       content: <CollectionTabContent selectedSubTab={...}>
    }
]}>
```

File: packages/compass-collection/src/components/workspace/workspace.jsx
Comment (mcasimir) [inline packages/compass-collection/src/components/workspace/workspace.jsx:189]: This may be a consequence of my lack of context here, but isn't there a bit of mix and match of responsibilities and knowledge split between this component and WorkspaceTabs right now, what is the idea behind both?

File: packages/compass-workspace/src/components/workspace-tabs.spec.tsx
Comment (mcasimir) [inline packages/compass-workspace/src/components/workspace-tabs.spec.tsx:95]: Why do we want this component to do this?

Comment (Anemy) [reply packages/compass-workspace/src/components/workspace-tabs.spec.tsx:95]: We want the tabs to contains things like `Document` or `Explain Plan` that indicate the contents of the tab.
This test checks that the tabs rendered by `workspace-tabs` contain the correct information passed through. It's definitely a bit of a coupled test as it pairs the rendering in `tab` with `workspace-tabs`, but I think that helps build some guarantee that things are rendered and passed through. Does that answer the question?

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (gribnoysup) [reply packages/compass-workspace/src/components/workspace-tabs.tsx:233]: Yeah, something like this would be even better, I agree

File: packages/compass-collection/src/components/workspace/workspace.jsx
Comment (Anemy) [reply packages/compass-collection/src/components/workspace/workspace.jsx:189]: What is the knowledge split here?
I think the idea is that `WorkspaceTabs` is a more straightforward rendering of the tabs for the workspace, and this `workspace.jsx` manages the rendering of the tabs and some of the global keyboard actions like creating a new tab with `ctrl/cmd + t`. If I'm not following your train of thought let's hop on a call?

File: packages/compass-workspace/src/components/tab.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/tab.tsx]: Thanks for catching! Here's how it looks now:
Screen Shot 2022-03-16 at 8 44 55 AM

File: packages/compass-workspace/src/components/workspace-tabs.spec.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/workspace-tabs.spec.tsx:95]: Would it be better to rename this `activeSubTabName` to `workspaceName` to be less collection/subtab centric in the naming?

File: packages/compass-collection/src/components/workspace/workspace.jsx
Comment (mcasimir) [reply packages/compass-collection/src/components/workspace/workspace.jsx:189]: I think I'm probably just confused by the fact that there is a workspace package and then this file which seems to control `WorkspaceTabs` in every aspect through its own store. But that's just temporary or?

File: packages/compass-workspace/src/components/workspace-tabs.spec.tsx
Comment (mcasimir) [reply packages/compass-workspace/src/components/workspace-tabs.spec.tsx:95]: Renaming it would definitely be better yeah. But I'll never be fond of this coupling here, in my opinion, `WorkspaceTabs` shouldn't know anything about sub-tabs, tabs content, and the fact that there is a concept of time-series and whatnot.

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/workspace-tabs.tsx]: Good thinking, I've updated to make most of the component more generic and accept an icon.
I think it might make sense to avoid moving the content rendering into this `compass-workspace` in this pr, I'll update it so that we're passing a render function from `compass-collection` as `compass-collection` has a good amount of logic tied around how the tabs are generated and even some ipc listeners for interacting with the tabs.

File: packages/compass-workspace/src/components/tab.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/tab.tsx]: Updated - added a `useRovingTabIndex` for focusing the tabs, and now only the selected/active tab will have `tabIndex={0}`. This `useRovingTabIndex` could live in `compass-components`, but since this is the only use place of it I haven't moved it there yet. Would you prefer I move it over now? Otherwise we can just keep an eye on it and move it over if we end up needing it elsewhere.

File: packages/compass-collection/src/components/workspace/workspace.jsx
Comment (Anemy) [reply packages/compass-collection/src/components/workspace/workspace.jsx:225]: Synced offline on how we can move towards a generic `compass-workspace` while addressing around the current collection centric logic in the `compass-collection` package. Down the line we'll use providers to supply the ipc handlers that are currently in this store. We can then manage listening to ipc events, like the menu actions from the main compass process as part of our react lifecycle. This'll probably clear up some bugs we currently have in Compass.

File: packages/compass-workspace/src/components/workspace-tabs.spec.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/workspace-tabs.spec.tsx]: Updated 👍  thanks for raising.

File: packages/compass-workspace/src/components/tab.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/tab.tsx]: Screen Shot 2022-03-16 at 10 45 40 PM

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (gribnoysup) [inline packages/compass-workspace/src/components/workspace-tabs.tsx:295]: This should be probably also be a prop for now and say something like "Collections" or "Collection Workspaces" for now, does this makes sense?

File: packages/compass-workspace/src/components/tab.tsx
Comment (gribnoysup) [reply packages/compass-workspace/src/components/tab.tsx]: I think if it's something that only used in one place it would make sense not to move it. Kind of a tangent, but still related: I do feel like now with the updated, more generic tabs interface that this component has this whole thing can be part of compass-components and not a separate package, would that make sense to you?

File: packages/compass-collection/src/components/workspace/workspace.jsx
Comment (mcasimir) [inline packages/compass-collection/src/components/workspace/workspace.jsx:241]: Why can't this be just `icon: <Elem>`?

Comment (Anemy) [reply packages/compass-collection/src/components/workspace/workspace.jsx:241]: It could, where would the styling be applied then? I'm cool with whatever here.

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/workspace-tabs.tsx]: Sounds good, created a ticket in the epic to move more of the workspace into the package and make the boundary between `Collection` view and `Workspace` view a bit more distinct: https://jira.mongodb.org/browse/COMPASS-5605

File: packages/compass-collection/src/components/workspace/workspace.jsx
Comment (mcasimir) [reply packages/compass-collection/src/components/workspace/workspace.jsx:242]: ah cause we need to apply some style to the icon, makes sense

File: packages/compass-workspace/src/components/tab.tsx
Comment (gribnoysup) [inline packages/compass-workspace/src/components/tab.tsx:269]: If we want to be so much in control of the props that are passed to the icon, wouldn't it make sense to only allow to pass valid leafygreen glyph names here and nothing else? With the current implementation I can break the layout easily by setting my own size or a class name after spreading the props and the props are not generic enough to be spreadable to anything, like if I use another icon component that doesn't support size exactly as expected here it wouldn't work

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (gribnoysup) [inline packages/compass-workspace/src/components/workspace-tabs.tsx:307]: I think this `ref={tabContainerRef}` will be overwritten by `tabContainerProps`, you sure it works as expected? Is it just not used anywhere?

**Review by gribnoysup — APPROVED** [review]
As I mentioned to you in person I do think that this generic component doesn't need a separate package and can be part of compass-components, maintaining separate package comes with an overhead and I don't see any benefit of having it separated. I don't have a strong enough opinion here though, so I will not block this work on this, your call if you want to keep it here or move it.

I also spotted a few small things, but nothing blocking too. Otherwise these new tabs look and work awesome!

File: packages/compass-components/src/compass-font-sizes.ts
Comment (Anemy) [reply packages/compass-components/src/compass-font-sizes.ts:2]: Yes it exists there. We're not using this `defaultFontSize` in many places in Compass. I can open up a pr replacing those and removing it, but it's probably not worth doing here as its a bit unrelated. How do you feel about the addition of the `smallFontSize` ? Is there another way we might want to do this? We could add some typography components to avoid a constant here.

File: packages/compass-workspace/src/components/workspace-tabs.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/workspace-tabs.tsx]: Ah thanks! Missed removing it when cleaning up.

File: packages/compass-workspace/src/components/tab.tsx
Comment (Anemy) [reply packages/compass-workspace/src/components/tab.tsx]: Sounds good, updated it to take a `iconGlyph` string with the `glyph` type from leafygreen.

File: packages/compass-components/src/compass-font-sizes.ts
Comment (gribnoysup) [reply packages/compass-components/src/compass-font-sizes.ts:2]: This font size is completely out of leafygreen design system, right? Is the difference that noticeable that we can't use 14px here? If it is then I guess it's fine, but 1px sounds like something that can be changed to a slightly bigger font without major issues. Anyway, totally not something that should block this PR from landing

Comment (Anemy) [reply packages/compass-components/src/compass-font-sizes.ts:2]: I think so yes. Claudia mentioned the 13 px size explicitly but 14px I'm thinking could work, I'll check with her. In the meantime I think I'll merge and we can update those fonts independently (and remove the 14px).

**Pull request merged**
